server/mailet/integration-testing/src/test/java/org/apache/james/mailets/RemoteDeliveryErrorTest.java

```diff
@@ -0,0 +1,196 @@
+    private static final String ALWAYS_421_RCPT_BEHAVIOR = "[{" +
+        "  \"condition\": {\"operator\":\"matchAll\"}," +
+        "  \"response\": {\"code\":421, \"message\":\"mock response\", \"rejected\":true}," +
+        "  \"command\": \"RCPT TO\"" +
+        "}]";
+    private static final String ALWAYS_421_FROM_BEHAVIOR = "[{" +
+        "  \"condition\": {\"operator\":\"matchAll\"}," +
+        "  \"response\": {\"code\":421, \"message\":\"mock response\", \"rejected\":true}," +
+        "  \"command\": \"MAIL FROM\"" +
+        "}]";
+    private static final String ALWAYS_421_DATA_BEHAVIOR = "[{" +
+        "  \"condition\": {\"operator\":\"matchAll\"}," +
+        "  \"response\": {\"code\":421, \"message\":\"mock response\", \"rejected\":true}," +
+        "  \"command\": \"DATA\"" +
+        "}]";
+    private static final String BOUNCE_MESSAGE = "Hi. This is the James mail server at localhost.\n" +
+        "I'm afraid I wasn't able to deliver your message to the following addresses.\n" +
+        "This is a permanent error; I've given up. Sorry it didn't work out. Below\n" +
+        "I include the list of recipients and the reason why I was unable to deliver\n" +
+        "your message.";
+
+    @Rule
+    public TemporaryFolder temporaryFolder = new TemporaryFolder();
+    @Rule
+    public IMAPMessageReader imapMessageReader = new IMAPMessageReader();
+    @Rule
+    public SMTPMessageSender messageSender = new SMTPMessageSender(DEFAULT_DOMAIN);
+    @ClassRule
+    public static DockerContainer mockSmtp = DockerContainer.fromName("chibenwa/mock-smtp-server")
+        .withLogConsumer(outputFrame -> System.out.println(outputFrame.getUtf8String()));
+
+    private TemporaryJamesServer jamesServer;
+
+    @Before
+    public void setUp() throws Exception {
+        InMemoryDNSService inMemoryDNSService = new InMemoryDNSService()
+            .registerMxRecord(DEFAULT_DOMAIN, LOCALHOST_IP)
+            .registerMxRecord(ANOTHER_DOMAIN, mockSmtp.getContainerIp());
+
+        jamesServer = TemporaryJamesServer.builder()
+            .withBase(SMTP_AND_IMAP_MODULE)
+            .withOverrides(binder -> binder.bind(DNSService.class).toInstance(inMemoryDNSService))
+            .withMailetContainer(MailetContainer.builder()
+                .putProcessor(CommonProcessors.simpleRoot())
+                .putProcessor(CommonProcessors.error())
+                .putProcessor(directResolutionTransport())
+                .putProcessor(CommonProcessors.bounces()))
+            .build(temporaryFolder.newFolder());
+
+        jamesServer.getProbe(DataProbeImpl.class)
+            .fluent()
+            .addDomain(DEFAULT_DOMAIN)
+            .addUser(FROM, PASSWORD);
+    }
+
+    @After
+    public void tearDown() {
+        jamesServer.shutdown();
+    }
+
+    @Test
+    public void remoteDeliveryShouldBounceWhenAlwaysRCPT421() throws Exception {
+        given(requestSpecification())
+            .body(ALWAYS_421_RCPT_BEHAVIOR)
```

Comment:
could be cool if we have a MockSMTPClient :)

Reply:
IMO the benefit do not overcome the complexity costs. IMO raw calls to restassured are enough.